lib/validate.js

```diff
@@ -12,25 +12,25 @@ function validate (state, doc) {
   .then(function () {
     return state.validate(doc)
   })
-  .catch(function (error) {
-    var err = new Error()
+  .catch(function (rejectValue) {
+    var error = new Error()
 
-    if (!(error instanceof Error)) {
-      if (typeof error === 'string') {
-        err.message = error
+    if (!(rejectValue instanceof Error)) {
+      if (typeof rejectValue === 'string') {
+        error.message = rejectValue
       } else {
-        err.message = 'check error value for more details'
-        err.value = error
+        error.message = 'check error value for more details'
+        error.value = rejectValue
       }
     } else {
-      if (!error.message) {
-        err.message = 'your doc(s) failed validation'
+      if (!rejectValue.message) {
+        error.message = 'your doc(s) failed validation'
       } else {
-        err.message = error.message
+        error.message = rejectValue.message
       }
     }
```

Comment:
I would sup around the if / else statements. Try avoiding negations. E.g. do `if (rejectValue instanceof Error) {` instead of `if (!(rejectValue instanceof Error)) {` and `if (rejectValue.message) {` instead of `if (!rejectValue.message) {`

I was also thinking that maybe people will throw errors with custom data on it, e.g. think a request error that fails with a `.status` or `.statusCode` property. These would get lost in this code. 

What we could do is that if `rejectValue` is `instanceof Error`, we can directly edit that errors `.name` and defaults its `.message` property to `your doc(s) failed validation`?

🤔

Reply:
I did some local edits and turns out the error we throw is still represented as `Error: <message>` rather than `ValidationError: <message>`
Now, I was hoping that it would work as I was avoiding creating a custom Error as it requires the use of `this`.
I know hoodie tries to avoid using `this` in the code, so what do you propose we do about this?

Reply:
^ @gr2m

Reply:
This is odd 🤔 It should :tm: be possible to change an error’s name:

```js
const error = new Error('test')
error.name = 'ValidationError'
error.toString()
// ValidationError: test
```

I don’t see how we would need to use `this` here? Sorry I must miss something obvious :(

Reply:
Try throwing this error. You'll see the difference :)

Reply:
wow I didn’t know you are right

```js
const error = new Error('test')
error.name = 'ValidationError'
throw error
// Unncaught Error: test
```

But try this

```js
error = new Error('test')
error.name = 'ValidationError'

Promise.resolve()

.then(() => {throw error})

.catch((error) => console.log(error.toString()))
// logs ValidationError: test
```

So my guess is it should work?

Reply:
Need a slight tweak to what you propose...

```js
error = new Error('test')
error.name = 'ValidationError'

Promise.resolve()

.then(() => {throw error})

.catch((error) => {
  console.error(error.toString())
  return Promise.reject()
})
```

Let me know if we can finalize this, @gr2m

Reply:
My code was just an example to showcase that the custom error name works. I think we should be good now, if you could finalize it and test the custom error name in a test, that’d be great 👍